### ISSUE TITLE
JSON-like serialization on .dict()

### ISSUE THREAD

**Issue by wozniakty** [pydantic/pydantic#1409]
# Feature Request

pydantic version: 1.5
pydantic compiled: True
install path: C:\Work\repos\data-purge\.venv\Lib\site-packages\pydantic
python version: 3.8.1 (tags/v3.8.1:1b293b6, Dec 18 2019, 23:11:46) [MSC v.1916 64 bit (AMD64)]
                     platform: Windows-10-10.0.18362-SP0
optional deps. installed: []

I'm using a framework (lambda) that expects the lambdas to return a simple dict to be passed along to another step.  This is done through marshaling that I don't really have a lot of access to modifying.  I'm using pydantic models for all of my de/serialization, and it had been working great to just say something to the effect of `return result.dict()`

However, now that I've added a datetime field to the model, I get a failure on: `Unable to marshal response: Object of type datetime is not JSON serializable`

Now, I know this is a pretty standard python issue with anything Json related.  But given that pydantic is such a great library for managing all this stuff, and it's not uncommon to have _another_ library expect to receive dicts, not json strings, I think it would be nice if we could have some way for `.dict()` to return a serializable format by default.

Essentially, I want to still have my pydantic model define the fields as a datetime, but I want the `.dict` step to convert those into a string format, rather than the datetime objects themselves.  Is this at all possible/reasonable?

```py
In [1]: from pydantic import BaseModel

In [2]: from datetime import datetime

In [3]: class Thing(BaseModel):
   ...:     x: datetime
   ...: 
In [5]: a = Thing(x="2020-01-01T00:00:00")

In [6]: a.dict()
Out[6]: {'x': datetime.datetime(2020, 1, 1, 0, 0)}

# preferred output:
Out[6]: {"x": "2020-01-01T00:00:00"}
...
```

Is this already possible?  Could this be added in some kind of way?  This isn't the first time I've hit this scenario, and I can't find an elegant way around it.  If only python actually was reasonable about datetimes, I wouldn't expect pydantic or something else to solve it :/

**Comment by StephenBrown2** [comment]
The thing is, dicts are intended to have python objects in them. All pydantic models do have a [`.json()` function](https://pydantic-docs.helpmanual.io/usage/exporting_models/#modeljson) which will automatically serialize common items into something json-serializable, and you can use the [`json_encoders` config dict](https://pydantic-docs.helpmanual.io/usage/exporting_models/#json_encoders) to customize the format if you wish, but by default, this is what you get with `.json()`:
```py
In [1]: from pydantic import BaseModel

In [2]: from datetime import datetime

In [3]: class Thing(BaseModel):
   ...:     x: datetime
   ...: 
In [5]: a = Thing(x="2020-01-01T00:00:00")

In [6]: a.json()
Out[6]: '{"x": "2020-01-01T00:00:00"}'
```
Which means that you can round-trip it back to a dict if you wish with:
```py
In [7]: import json

In [8]: json.loads(a.json())
Out[8]: {'x': '2020-01-01T00:00:00'}
```
Which is how I handled things before `exclude_none` was a thing.

**Comment by wozniakty** [comment]
I can _probably_ accept that this is just a limitation conceptually in the way some of the underlying tools are defined.  But I will say that conceptually, in principle, objects go through a transition of 4 states: `Model > Dict w/ python objects > dict of serializable types > string of dict`

It's unfortunate to me that it's not possible to actually step to the third state without doing something like skipping to step 4 and rolling back to step 3 by way of `json.loads`

**Comment by StephenBrown2** [comment]
You can also define a custom type that validates a datetime and returns the string version, something like:

```py
In [1]: from datetime import datetime
   ...: from pydantic import BaseModel
   ...: from pydantic.datetime_parse import parse_datetime
   ...:
   ...:
   ...: class StringDate(datetime):
   ...:     @classmethod
   ...:     def __get_validators__(cls):
   ...:         yield parse_datetime
   ...:         yield cls.validate
   ...:
   ...:     @classmethod
   ...:     def validate(cls, v: datetime):
   ...:         return v.isoformat()
   ...:
   ...: class Thing(BaseModel):
   ...:     x: StringDate
   ...:
   ...: a = Thing(x="2020-01-01T00:00:00")
   ...:
   ...: a.dict()
Out[1]: {'x': '2020-01-01T00:00:00'}
```

**Comment by seansfkelley** [comment]
I've had a similar problem as well. My use-case is that I'm writing an HTTP handler that traffics in Pydantic models, but in a framework that has baked-in opinions about how it wants to do things like encode datetimes that I don't agree with but have to work with.

The current pattern I use is something like `json.loads(model.json())`, which yields an object that's unambiguous dicts/lists/strs etc. without any fancy types, and then the HTTP framework happily encodes that object into a JSON string (again) without doing anything surprising to my datetimes, which are now in string form. (If I were to just pass `model.json()`, it would escape the already-jSON string and produce nonsense... argh!)

Relatedly, there are cases where I want to `dict()` something to pass to other Python code to continue processing, but also take advantage of `json_encoders` to convert some string-wrapping newtype (e.g. Mongo's `bson.ObjectId`) into a normal object (`str`).

In short, it would be nice if there were a method halfway between `dict` and `json` that basically produced vanilla Python objects that were unambiguously convertible to JSON types but without the performance penalty of round-tripping through a JSON string, that is, semantically equivalent to `json.loads(model.json())`. Strawman proposal: `model.json(objects=True)`.

**Comment by wozniakty** [comment]
Another notable awkwardness (again, not a _problem_ but just a suggestion of why it feels inconsistent) is that if you have a sub-object that's a BaseModel, it's not like `.dict()` returns that as a BaseModel.  It dict-ifies every pydantic model in the tree, but keeps pythony objects for everything else.

So one might argue that if `.dict()` serializes the BaseModel all the way down the tree, there might be a way to do the same for other types.

**Comment by eruvanos** [comment]
We had the same issue, to write into a dynamodb.

For this we use:

```
from fastapi.encoders import jsonable_encoder

class Thing(BaseModel):
    x: datetime

a = Thing(x="2020-01-01T00:00:00")

json_compatible_item_data = jsonable_encoder(a)
```

https://fastapi.tiangolo.com/tutorial/encoder/

**Comment by samuelcolvin** [comment]
This is not a new idea @tiangolo submitted a PR for this way back in 2018 #317. I refused that PR, but I think we should reconsider it.

The reasons I refused that PR are explained in https://github.com/samuelcolvin/pydantic/pull/317#issuecomment-443689941 but it comes down to basically: "`ujson.loads(m.dict()` will likely be faster" and "not everyone uses JSON and people will want  different behaviour". I also want to be able to reuse the serialization logic [in `pydantic_encoder `](https://github.com/samuelcolvin/pydantic/blob/8e00077cc4151a666a11527292d45e36e90067f4/pydantic/json.py#L45).

I think we should do the following:
* add a `serialize` kwarg to `dict()` which has type `Union[bool, Callable] = False` - when `True` the model is returned as a dict with only JSON types (this will use `pydantic_encoder`), when `False` (the default) the current behaviour is maintained, when a callable that callable will be called on all objects that aren't valid JSON to allow you to customise how things are serialised.
* in v2, rename `pydantic/json.py` to `pydantic/serialisation.py` as it's not just related to JSON encoding.

**Comment by tiangolo** [comment]
I think it would be great to have some form of this :tada: 

Just a note about the implementation, it could make sense to have an external function to do the serialization instead of a class method (or additional to?), that way it could support serializing an list of type `List[SomeModel]` and not only a sub-class of `BaseModel` (e.g. `SomeModel`).

Being able to serialize a list of models would still be needed either way for fields of a type like that, e.g.:

```Python
from pydantic import BaseModel

class SubModel(BaseModel):
    name: str

class Model(BaseModel):
    sequence_field: List[SubModel]
```

Here `Model.sequence_field` would need that serialization of list of models. But being able to use it outside of fields could be potentially useful and I think it would probably involve a similar effort.

**Comment by cmlccie** [comment]
Adding the `serialize` kwarg to `dict()` as described above, will be a significant step forward in clarity and performance for a use case that I frequently encounter - serializing to YAML. 🙂

It makes sense that pydantic should not attempt to include output serialization for *all* data formats (not reasonable or sustainable). Exposing pydantic's capability to serialize complex data types into *basic* Python types provides excellent utility to everyone desiring to serialize to some *other* output format.

**Comment by wozniakty** [comment]
@samuelcolvin excited to see this gaining some traction!  If I might make a suggestion:

I think I would weak-prefer (as in, fine if it doesn't happen, but just stating a preference) a pattern of `.dict(), .serialize(), .json()` over a `.dict(serialize=), .json()` as it's more functionally accurate that a pydantic model is essentially passing between three states.  `BaseModel -> Dict w/ types specified by model -> Dict with serializeable types -> json string`.  These states seem best represented by 3 independent functions, IMO.

Along with that, I might recommend the signature for `serialize` be something like:
```
def serialize(self, type_converters: Dict[Union[Type, Literal['_']], Callable[[Type], Any] = default_json_encoder_func) -> Dict[str, Any]:
    # pseudocode
    for field in fields:
        if type(field) in type_converters:
            converted_type = type_converters[type(field)](field)
```
Where instead of specifying a function that converts a whole object, we might say that someone wants to put a specific set of conversions on particular types.  They could of course say `{ "_": convert_all_types()}` which is implied to be the converter for any types not explicitly specified (the `_` matching the default pattern matching indicator in [PEP-622](https://www.python.org/dev/peps/pep-0622/))

Now, some obvious caveats:
- Perf concerns may make this kind of implementation more complex of a refactor, I'm uncertain as I haven't dug fully into the `.dict` and `.json` code yet.
- names open to debate, obviously.  You could make the argument that `serialize` implies something more than just a python dict return.
- matching on type might be a perf concern as well, I'm not super confident in my understanding of python perf nuances at that level, so maybe the serialize function is best 🤷

**Comment by wozniakty** [comment]
Update: I'm working on a PR for this, now I understand `pydantic_encoder` better and see why the dict structure isn't really relevant.  Will go with accepting a callable.

**Comment by samuelcolvin** [comment]
I don't agree about adding another method, see #1001

**Comment by Sheshtawy** [comment]
Is anyone working on this? I would like to work on it if nobody is

**Comment by wozniakty** [comment]
@Sheshtawy I have not had time to consider a new approach since my closed PRs, so go ahead.

**Comment by Sheshtawy** [comment]
@samuelcolvin In reference to your comments 
Here:
> We'll have to take some care over how we implement serialize=True to maintain good performance.

And here:  https://github.com/samuelcolvin/pydantic/pull/1986#issuecomment-716188634

What concerns do you have regarding implementation here? What other things would like to see addressed in a PR addressing this feature? I know performance is an important priority for the library, but what does that imply in terms of implementation?

**Comment by uselessscat** [comment]
I have a similar problem. I want to create a sqlalchemy model from a pydantic schema

```
MyModel(**pydantic_schema.dict())
```

in my model a field is defined as json

```
from sqlalchemy.schema import Column
from sqlalchemy.types import Integer, JSON
from app.database import BaseModel # declarative base

class MyModel(BaseModel):
    id = Column(Integer, primary_key=True, autoincrement=True)
    json_field = Column(JSON, nullable=False)
```

The pydantic schema is defined like this:

```
from datetime import datetime
from pydantic import BaseModel, Field

class SomeJsonField:
    # This recursive model is to force some json structure
    string_field: str = Field(...)
    datetime_field: datetime = Field(...)

class MySchema(BaseModel)
    id: int = Field(...)
    json_field: SomeJsonField = Field(...)
```

This raises an exception because Sqlalchemy tries to serialize datetime objects from the `.dict()` results.
It would be great if I can (in some way) specify what level of serialization I want to do.

just like json but for dict serializations:

```
class WithCustomEncoders(BaseModel):
    dt: datetime

    class Config:
        json_encoders = {
            datetime: lambda v: v.timestamp(),
        }
        dict_encoders = {
            datetime: lambda v: v.timestamp(),
        }
```

Is there a way to achieve what I want?

**Comment by couling** [comment]
If I read correctly, there's still no working solution for this.  Until one is found, I wonder if the documentation could draw more attention to `pydantic.json.pydantic_encoder`.  As it stands, it's not mentioned under [`model.json()`](https://pydantic-docs.helpmanual.io/usage/exporting_models/#modeljson) or [`model.dict()`](https://pydantic-docs.helpmanual.io/usage/exporting_models/#modeldict).  I found myself reading pydantic source code before I tripped over it.

My issue (subtally different from the OP's) is I'm storing values in sqlalchemy json fields.  I can't help the fact that sqlalchemy want's control over the conversion between dict and raw json.  So I must go via `model.dict()` not `model.json()`.

I've come up with two work arounds.  The second might help the OP and both may be useful in other similar issues.

-----

## Work around 1

**If you have control over the json encoder** (which the OP does not) then you can pass `default=pydantic.json.pydantic_encoder` into `json.dumps`.  Eg: for sqlalchemy I was able to:

```python
import json
import pydantic.json
import sqlalchemy

def _custom_json_serializer(*args, **kwargs) -> str:
    return json.dumps(*args, default=pydantic.json.pydantic_encoder, **kwargs)

engine = sqlalchemy.create_engine(connection_string, json_serializer=_custom_json_serializer)
# Use this engine as normal.
```

-----

## Work around 2

**If you don't have control over the json encoder** you can create a method which will re-encode your dictionary ready for json encoding.

```python
import pydantic.json

# https://github.com/python/cpython/blob/7b21108445969398f6d1db9234fc0fe727565d2e/Lib/json/encoder.py#L78
JSONABLE_TYPES = (dict, list, tuple, str, int, float, bool, type(None))

def to_jsonable_dict(obj):
    if isinstance(obj, dict):
        return {key: to_jsonable_dict(value) for key, value in obj.items()}
    elif isinstance(obj, list):
        return [to_jsonable_dict(value) for value in obj]
    elif isinstance(obj, tuple):
        return tuple(to_jsonable_dict(value) for value in obj)
    elif isinstance(obj, JSONABLE_TYPES):
        return obj
    return pydantic.json.pydantic_encoder(obj)
```
Then `json.dumps(to_jsonable_dict(model.dict()))` should succeed where `json.dumps(model.dict())` would fail.  Though neither solution takes into account [json encoders configured on the object](https://pydantic-docs.helpmanual.io/usage/exporting_models/#json_encoders).  For that, a solution is still needed in pydantic itself.

**Comment by radusuciu** [comment]
I also would like this feature, or really any features that provide more control over serialization. My use case is parsing a custom INI-like configuration file. When writing it back out again into the custom format, it's easier to work with basic data types. I'm currently making use of `use_enum_values = True` to get part of the way there and for nested models I've overloaded `dict` to return a string representation of the value (please let me know if this is a terrible idea - I'm new to Pydantic). For my use case it almost seems like the best way would be to have per-field serializers that can be used as decorators -- basically the same as `@validator` currently works.

**Comment by havardthom** [comment]
+1 on this feature, I'm using FastAPI with pydantic and neo4j. Querying neo4j with their python driver expects parameters to be passed as dictionary and they do not support data types such as UUID or ObjectId, so I have to manually cast them after using `.dict()` or use some other hacky workaround. Would love to be able to pass a custom encoder function to `.dict()` to serialize certain data types.

**Comment by frediy** [comment]
Adding in my +1 for this. Using this to hit gcloud firestore where you pass the docs to be stored as dicts. Currently breaks due to enums, and I'm using the `json.loads(model.json())` thing atm.

**Comment by Rudedog9d** [comment]
Another use case is for PyMongo, which accepts a dict for inserting into a database. Currently using this:

```py
MongoClient(..., fallback_encoder=CustomBaseModel.__json_encoder__))
```

**Comment by ChristianMertes** [comment]
To add yet another use case: [Anvil](https://anvil.works/) server methods.

**Comment by bafonso** [comment]
Adding my +1 too, this would be extremely useful

**Comment by antoinedelanglois** [comment]
+1 for me too!

**Comment by tkaraouzene** [comment]
+1 for me too

**Comment by DavidMChan** [comment]
Hi @samuelcolvin  - I was looking at submitting a PR for this following the proposed api in https://github.com/samuelcolvin/pydantic/issues/1409#issuecomment-650116116, but it seems like there are a couple of edge cases which would need to be resolved between standard `.dict()` and standard `.json()` behavior. 

The two I found are:

```python
def test_serialize_custom_root():
    class Model(BaseModel):
        __root__: List[str]
    assert Model(__root__=['a', 'b']).dict(serialize=True) == ["a", "b"]
```

Where `.dict()` produces `{'__root__': ["a","b"]}` and `.json()` produces `["a","b"]`. I feel like this should be consistent between the two - but perhaps there's some underlying reason for the two producing a different output. In the case of the `serialize=True` option, this would have to be resolved in one direction or another.

The second edge case that I found was:

```python
def test_json_nested_serialize_models():
    class Phone(BaseModel):
        manufacturer: str
        number: int

    class User(BaseModel):
        name: str
        SSN: int
        birthday: datetime.datetime
        phone: Phone
        friend: Optional['User'] = None  # noqa: F821  # https://github.com/PyCQA/pyflakes/issues/567

        class Config:
            json_encoders = {
                datetime.datetime: lambda v: v.timestamp(),
                Phone: lambda v: v.number if v else None,
                'User': lambda v: v.SSN,
            }

    User.update_forward_refs()
    galaxy = Phone(manufacturer='Samsung', number=18007267864)
    pumbaa = User(
        name='Pumbaa', SSN=234, birthday=datetime.datetime(1993, 5, 15, tzinfo=datetime.timezone.utc), phone=galaxy
    )
    assert (
        pumbaa.json(models_as_dict=False)
        == '{"name": "Pumbaa", "SSN": 234, "birthday": 737424000.0, "phone": 18007267864, "friend": null}'
    )
```

Where  `.dict()` doesn't require resolving the custom json_encoder (and instead does this resolution through include/exclude), while `.json()` will resolve the custom encoders, and produce a cut-down format. 

If you want to support `.json()` like behavior for `dict(serialize=True)`, then it requires a relatively complex re-write of the way that `.dict()` encoding is handled. If we take the .dict() path, then the code for the PR is actually relatively relaxed, and very similar to what was proposed in #1986 (and I think we can probably keep the changes out of the `json.py` file, and limited mostly to the _get_value function). 

I didn't find anything else using the tests from both `.dict()` and `.json()`, but I would appreciate a mind-dump of any other considerations for an implementation, which would be really helpful when moving towards a solution on this issue.

A sketch of my changes so far are visible in https://github.com/DavidMChan/pydantic/tree/davidmchan/serialize_refactor
(However this is a pretty rough WIP, and includes some extra code from several different implementation possibilities).

**Comment by deanq** [comment]
I'm surprised no one has put up this solution. I'm using pydantic for boto3 dynamodb interface. This is how I deal with `instance.dict()` to use string datetime.
```python
from pydantic import BaseModel
from datetime import datetime

class Model(BaseModel):
    timestamp: datetime

    def dict(self, **kwargs):
        output = super().dict(**kwargs)
        for k,v in output.items():
            if isinstance(v, datetime):
                # datetime: "2022-04-16T06:43:56+00:00"
                output[k] = v.isoformat()
        return output
```

**Comment by IsmaelMT** [comment]
> I'm surprised no one has put up this solution. I'm using pydantic for boto3 dynamodb interface. This is how I deal with `instance.dict()` to use string datetime.
> 
> ```python
> from pydantic import BaseModel
> from datetime import datetime
> 
> class Model(BaseModel):
>     timestamp: datetime
> 
>     def dict(self, **kwargs):
>         output = super().dict(**kwargs)
>         for k,v in output.items():
>             if isinstance(v, datetime):
>                 # datetime: "2022-04-16T06:43:56+00:00"
>                 output[k] = v.isoformat()
>         return output
> ```

I think that's a good idea for certain cases, it won't work for nested fields though. You'd need to do something like the following: 

```
    def _serialize(self, output: dict):
        for k, v in output.items():
            if isinstance(v, datetime):
                # datetime: "2022-04-16T06:43:56+00:00"
                output[k] = v.isoformat()
            if isinstance(v, dict):
                self._serialize(v)

    def dict(self, **kwargs):
        output = super().dict(**kwargs)
        self._serialize(output)

        return output
``
`

However, that adds an extra recursive pass apart from the one that's done in [_get_value](https://github.com/samuelcolvin/pydantic/blob/master/pydantic/main.py#L704), which is not ideal when you are looking at performance.

**Comment by Cataldir** [comment]
Anyone working on this? I might be able to help if there is someone, or do it otherwise.

**Comment by samuelcolvin** [comment]
This needs to be implemented in rust as part of pydantic-core, but first I
need to develop a full plan for serialisation in general.

Best to hold off for now.

**Comment by michaelkariv** [comment]
+1, waiting for the solution

**Comment by delatars** [comment]
Work with nested models.
```python
from pydantic import BaseModel
from datetime import datetime


class MyBaseModel(BaseModel):

    _encoders_by_type = {datetime: lambda dt: dt.isoformat(timespec='seconds')}

    def _iter(self, **kwargs):
        for key, value in super()._iter(**kwargs):
            yield key, self._encoders_by_type.get(type(value), lambda v: v)(value)


class SubModel(MyBaseModel):
    timestamp: datetime


class Model(MyBaseModel):
    timestamp: datetime
    sub_model: SubModel


m = Model(timestamp=datetime.now(), sub_model=SubModel(timestamp=datetime.now()))
m.dict()
# {'timestamp': '2023-01-10T02:11:08','sub_model': {'timestamp': '2023-01-10T02:11:08'}}
```

**Comment by harenbrs** [comment]
> Work with nested models.
> 
> ```python
> from pydantic import BaseModel
> from datetime import datetime
> 
> 
> class MyBaseModel(BaseModel):
> 
>     _encoders_by_type = {datetime: lambda dt: dt.isoformat(timespec='seconds')}
> 
>     def _iter(self, **kwargs):
>         for key, value in super()._iter(**kwargs):
>             yield key, self._encoders_by_type.get(type(value), lambda v: v)(value)
> 
> 
> class SubModel(MyBaseModel):
>     timestamp: datetime
> 
> 
> class Model(MyBaseModel):
>     timestamp: datetime
>     sub_model: SubModel
> 
> 
> m = Model(timestamp=datetime.now(), sub_model=SubModel(timestamp=datetime.now()))
> m.dict()
> # {'timestamp': '2023-01-10T02:11:08','sub_model': {'timestamp': '2023-01-10T02:11:08'}}
> ```

That's my favourite workaround so far. Also, if you already have custom `json_encoders` defined in the model config, you can do this instead:

```python 
    def _iter(self, **kwargs):
        for key, value in super()._iter(**kwargs):
            yield key, self.__config__.json_encoders.get(type(value), lambda v: v)(value)
```

then you don't need to define `_encoders_by_type`.

**Comment by JoaoPaes-at-Dynamox** [comment]
Two years now and still an issue... Come on...

**Comment by blakeNaccarato** [comment]
Of course it would be nice to have this resolved, but the project owner noted "best to hold off for now" above, as this needs a total rethink in their efforts in v2 and the Rust core. Any lag in issue resolution can be attributed to the substantial v2 effort. This issue is currently in the [v2 milestone](https://github.com/pydantic/pydantic/milestone/2), so it's being looked at. So the solution to this (and the effort to solve it) isn't likely to be bolted onto v1 when it's going to be handled fundamentally differently in v2.

See the rationale for this in the [Pydantic v2 Plan](https://docs.pydantic.dev/blog/pydantic-v2/#improvements-to-dumpingserializationexport). It's best to avoid noise on this issue and look forward to v2. In the meantime, workarounds for JSON-like serialization are presented in the comments above.

<details><summary>Author's note on "Improvements to Dumping/Serialization/Export"</summary>

> There has long been a debate about how to handle converting data when extracting it from a model.
> One of the features people have long requested is the ability to convert data to JSON compliant types while
> converting a model to a dict.
> 
> My plan is to move data export into pydantic-core, with that, one implementation can support all export modes without
> compromising (and hopefully significantly improving) performance.
> 
> I see four different export/serialisation scenarios:
> 
> 1. Extracting the field values of a model with no conversion, effectively `model.__dict__` but with the current filtering
>    logic provided by `.dict()`
> 2. Extracting the field values of a model recursively (effectively what `.dict()` does now) - sub-models are converted to
>    dicts, but other fields remain unchanged.
> 3. Extracting data and converting at the same time (e.g. to JSON compliant types)
> 4. Serialising data straight to JSON
> 
> I think all 4 modes can be supported in a single implementation, with a kind of "3.5" mode where a python function
> is used to convert the data as the user wishes.
> 
> The current `include` and `exclude` logic is extremely complicated, but hopefully it won't be too hard to
> translate it to Rust.
> 
> We should also add support for `validate_alias` and `dump_alias` as well as the standard `alias`
> to allow for customising field keys.

</details>

**Comment by samuelcolvin** [comment]
Thanks so much @blakeNaccarato for the adult response.

This is actually already implemented in pydantic-core and works on `main` now.

Example:

```py
from datetime import datetime
from pydantic import BaseModel


class MyModel(BaseModel):
    things: set[str]
    when: datetime


m = MyModel(things={'a', 'b'}, when=datetime.now())
print(m.model_dump(mode='json'))
"""
{
    'things': ['b', 'a'],
    'when': '2023-02-09T10:54:50.11279',
}
"""
```

Sorry for the wait, closing this.

We're working as fast as we can to get V2 released.

**Comment by jwilges** [comment]
Agree, @blakeNaccarato's response sums up what I was seeing/thinking after reading through all of the old threads and responses from @samuelcolvin. If anyone takes the time to look back and read the full context, you'll see @samuelcolvin's just trying to find the balance of getting this right rather than shoveling out a feature with compromises that might stir regret. And a few very workable workarounds were presented along the way, moreso than what I've seen sometimes for commercial libraries like boto3. This library is used by a lot of people... no sense in half-baking a feature like this one only to have to deprecate an interface later. And when you look at the V2 goals, getting this right in a generic way is on the roadmap.

Hey @samuelcolvin, regarding this line:

> This is actually already implemented in pydantic-core and works on main now.

That sort of blew my mind and got me stoked to check out what ideas "made the cut" to at least hit `main`.

I quickly checked the latest 1.10.6 release and didn't see it (not surprised, it's recent), then dug to find the most relevant bits and it looks like your snippet above was enabled with your [adopting pydantic-core serialisation (#4978)](https://github.com/pydantic/pydantic/pull/4978) PR specifically around [these main.py changes](https://github.com/pydantic/pydantic/commit/f4e8f71ab23a427662f8ac67ba04e453ece56cf3#diff-8708af8d26c26644976866f974e0b810b0940ede43ff487e457057d5a1f07999R228) that added the `mode` argument.

To give it a spin in a local project I just temporarily pip installed `'git+https://github.com/pydantic/pydantic.git@main'` but I thought I'd check to see if you foresee that changeset making the cut for a 1.10.x release later or is it something that might not hit the 1.10 series? Checking as I'm sure I'm not the only one following along who would like to keep an eye out and update a few pinned requirements files if it releases prior to 2.x.

Thanks again for the work on this project, for years now it has struck a nice balance of cutting edge and reliable, chopping out the super repetitive bits around serialization that felt dreadful to rewrite/maintain for every project.
